Handling flaky tests

For roughly the past month, I have collected data on flaky CI tests. The initial idea was to mark them as [flaky](https://github.com/box/flaky), but as per [pytest's docs on flaky tests](https://docs.pytest.org/en/stable/explanation/flaky.html), that should never be a long-term solution. Instead, tests should be (randomly) re-ordered, re-written for more atomic assertions, or split up into different groups to find the root cause of the flaky behavior and eliminate it. We will have to see when time permits this. For now, we could mark them as flaky to save us from re-running them manually. 
Here are the flaky tests of this repository I gathered so far:

# Flaky tests

## Auxiliary

Some error message are common and so long that they would render the table even more complex than it already is. They are included here and referenced by their names in the table below.

### Notebook cell timeout reticulate {#notebook-cell-timeout-reticulate}
```python
nbclient.exceptions.CellTimeoutError: A cell timed out while it was being executed, after 10 seconds.
The message was: Cell execution timed out.
Here is a preview of the cell contents:
-------------------
# Load reticulate, used to access the Python API from R
library(reticulate)

# Import ixmp and message_ix, just as in Python
ixmp <- import("ixmp")
-------------------
```

### Notebook cell timeout import packages {#notebook-cell-timeout-import-packages}
```python
nbclient.exceptions.CellTimeoutError: A cell timed out while it was being executed, after 10 seconds.
The message was: Cell execution timed out.
Here is a preview of the cell contents:
-------------------
# load required packages
import pandas as pd
import ixmp
-------------------
```

### Notebook cell timeout platform {#notebook-cell-timeout-platform}
```python
nbclient.exceptions.CellTimeoutError: A cell timed out while it was being executed, after 10 seconds.
The message was: Cell execution timed out.
Here is a preview of the cell contents:
-------------------
# launch the ix modeling platform using the local default database
mp <- ixmp$Platform()
-------------------
```

### Runtime error DB connection {#runtime-error-db-connection}
```python
RuntimeError: unhandled Java exception: 
Unable to obtain connection from database (jdbc:hsqldb:file:/private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/pytest-of-runner/pytest-0/test_multi_db_run0/mp2) for user 'ixmp': Database lock acquisition failure: lockFile: org.hsqldb.persist.LockFile@6486fbbd[file =/private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/pytest-of-runner/pytest-0/test_multi_db_run0/mp2.lck, exists=true, locked=false, valid=false, ] method: checkHeartbeat read: 2023-07-09 05:22:39 heartbeat - read: -862 ms.
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
SQL State  : S1000
Error Code : -451
Message    : Database lock acquisition failure: lockFile: org.hsqldb.persist.LockFile@6486fbbd[file =/private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/pytest-of-runner/pytest-0/test_multi_db_run0/mp2.lck, exists=true, locked=false, valid=false, ] method: checkHeartbeat read: 2023-07-09 05:22:39 heartbeat - read: -862 ms.
```

### DB connection cannot be closed {#db-connection-cannot-be-closed}
```python
AssertionError: assert 'Database connection could not be closed or was already closed' in ''
 +  where '' = CaptureResult(out='', err='').out
```

### DB connection message wrong {#db-connection-message-wrong}
```python
assert "connected to database 'jdbc:hsqldb:mem://ixmptest' (user: ixmp)..." in ''
 +  where '' = CaptureResult(out='', err='').out
```
Note that windows-latest-py3.10 had this additional information once:
```python
where '' = CaptureResult(out='', err="2023-07-03 12:58:22,174  INFO at.ac.iiasa.ixmp.Platform:182 - closed the connection to database 'jdbc:hsqldb:mem://ixmptest'\r\n2023-07-03 12:58:22,180  INFO at.ac.iiasa.ixmp.Platform:165 - Welcome to the IX modeling platform!\r\n2023-07-03 12:58:22,180  INFO at.ac.iiasa.ixmp.Platform:166 -  connected to database 'jdbc:hsqldb:mem://ixmptest' (user: ixmp)...\r\n").out
```



Names are shortened to `ixmp/tests` as the starting directory. Grouping the tests by names shows commonalities between them. All notebook cell timeout originate on macos, while ubuntu and windows only struggle with DB connections. Windows in particular seems to have trouble closing the DB connection either too early or not at all.


| Test name    | Error message | Runners (# of occurrences if > 1) |
|--------------|:-------------:|---------------------------:|
| test_tutorials.py:: test_R_transport_scenario |  [Notebook cell timeout reticulate](#notebook-cell-timeout-reticulate) | macos-latest-py3.7 (3)|
| test_tutorials.py:: test_R_transport | [Notebook cell timeout reticulate](#notebook-cell-timeout-reticulate) | macos-latest-py3.7 (5), macos-latest-py3.11 |
| test_tutorials.py:: test_py_transport | [Notebook cell timeout import packages](#notebook-cell-timeout-import-packages) | macos-latest-py3.7 |
| test_tutorials.py:: test_py_transport_scenario | [Notebook cell timeout import packages](#notebook-cell-timeout-import-packages) | macos-latest-py3.7 |
| test_tutorials.py:: test_R_transport_scenario | [Notebook cell timeout platform](#notebook-cell-timeout-platform) | macos-latest-py3.11 |
|------------------|------------------|------------------|
|------------------|------------------|------------------|
| test_access.py:: test_check_single_model_access | `ConnectionRefusedError: [Errno 61] Connection refused` | macos-latest-py3.7 |
|------------------|------------------|------------------|
|------------------|------------------|------------------|
| test_integration.py::test_multi_db_run | [Runtime error DB connection](#runtime-error-db-connection) | macos-latest-py3.8, macos-latest-py3.10 |
|------------------|------------------|------------------|
|------------------|------------------|------------------|
| backend/test_jdbc.py:: test_jvm_warn | `AssertionError: ResourceWarning("unclosed file <_io.BufferedReader name='/tmp/pytest-of-runner/pytest-0/test_read_excel_big0/output.xlsx'>") assert 1 == 0 where 1 = len(WarningsRecorder(record=True))` | ubuntu-latest-py3.8 | 
|------------------|------------------|------------------|
|------------------|------------------|------------------|
| backend/test_jdbc.py:: test_close | [DB connection cannot be closed](#db-connection-cannot-be-closed) | windows-latest-py3.7 (3), windows-latest-py3.8, windows-latest-py3.10 (2), windows-latest-py3.11 (2) |
| backend/test_jdbc.py:: test_connect_message | [DB connection message wrong](#db-connection-message-wrong) | windows-latest-py3.7 (3), windows-latest-py3.8, windows-latest-py3.10 (2), windows-latest-py3.11 (2) |

--- a/ixmp/backend/jdbc.py
+++ b/ixmp/backend/jdbc.py
@@ -385,7 +385,7 @@ class JDBCBackend(CachingBackend):
     def set_doc(self, domain, docs):
         dd = _domain_enum(domain)
         jdata = java.LinkedHashMap()
-        if type(docs) == dict:
+        if isinstance(docs, dict):
             docs = list(docs.items())
         for k, v in docs:
             jdata.put(str(k), str(v))

--- a/ixmp/tests/backend/test_jdbc.py
+++ b/ixmp/tests/backend/test_jdbc.py
@@ -1,5 +1,7 @@
 import gc
 import logging
+import os
+import platform
 from sys import getrefcount
 from typing import Tuple
 
@@ -14,9 +16,18 @@ from ixmp.backend.jdbc import DRIVER_CLASS, java
 from ixmp.testing import DATA, add_random_model_data, bool_param_id, make_dantzig
 from ixmp.testing.resource import memory_usage
 
+FLAKY = pytest.mark.flaky(
+    reruns=5,
+    rerun_delay=2,
+    condition="GITHUB_ACTIONS" in os.environ and platform.system() == "Darwin",
+    reason="Flaky; see iiasa/ixmp#489",
+)
+
+
 log = logging.getLogger(__name__)
 
 
+@FLAKY
 def test_jvm_warn(recwarn):
     """Test that no warnings are issued on JVM start-up.
 
@@ -38,6 +49,7 @@ def test_jvm_warn(recwarn):
         assert len(recwarn) == 0, recwarn.pop().message
 
 
+@FLAKY
 def test_close(test_mp_f, capfd):
     """Platform.close_db() doesn't throw needless exceptions."""
     # Use the session-scoped fixture to avoid affecting other tests in this file
@@ -176,6 +188,7 @@ def test_invalid_properties_file(test_data_path):
         ixmp.Platform(dbprops=test_data_path / "hsqldb.properties")
 
 
+@FLAKY
 def test_connect_message(capfd, caplog):
     msg = "connected to database 'jdbc:hsqldb:mem://ixmptest' (user: ixmp)..."
 

--- a/ixmp/tests/backend/test_jdbc.py
+++ b/ixmp/tests/backend/test_jdbc.py
@@ -16,18 +16,15 @@ from ixmp.backend.jdbc import DRIVER_CLASS, java
 from ixmp.testing import DATA, add_random_model_data, bool_param_id, make_dantzig
 from ixmp.testing.resource import memory_usage
 
-FLAKY = pytest.mark.flaky(
-    reruns=5,
-    rerun_delay=2,
-    condition="GITHUB_ACTIONS" in os.environ and platform.system() == "Darwin",
-    reason="Flaky; see iiasa/ixmp#489",
-)
-
-
 log = logging.getLogger(__name__)
 
 
-@FLAKY
+@pytest.mark.flaky(
+    reruns=5,
+    rerun_delay=2,
+    condition="GITHUB_ACTIONS" in os.environ and platform.system() == "Linux",
+    reason="Flaky; see iiasa/ixmp#489",
+)
 def test_jvm_warn(recwarn):
     """Test that no warnings are issued on JVM start-up.
 
@@ -49,7 +46,12 @@ def test_jvm_warn(recwarn):
         assert len(recwarn) == 0, recwarn.pop().message
 
 
-@FLAKY
+@pytest.mark.flaky(
+    reruns=5,
+    rerun_delay=2,
+    condition="GITHUB_ACTIONS" in os.environ and platform.system() == "Windows",
+    reason="Flaky; see iiasa/ixmp#489",
+)
 def test_close(test_mp_f, capfd):
     """Platform.close_db() doesn't throw needless exceptions."""
     # Use the session-scoped fixture to avoid affecting other tests in this file
@@ -188,7 +190,12 @@ def test_invalid_properties_file(test_data_path):
         ixmp.Platform(dbprops=test_data_path / "hsqldb.properties")
 
 
-@FLAKY
+@pytest.mark.flaky(
+    reruns=5,
+    rerun_delay=2,
+    condition="GITHUB_ACTIONS" in os.environ and platform.system() == "Windows",
+    reason="Flaky; see iiasa/ixmp#489",
+)
 def test_connect_message(capfd, caplog):
     msg = "connected to database 'jdbc:hsqldb:mem://ixmptest' (user: ixmp)..."
 

--- a/ixmp/tests/test_access.py
+++ b/ixmp/tests/test_access.py
@@ -1,4 +1,6 @@
 import logging
+import os
+import platform
 import sys
 from subprocess import Popen
 from time import sleep
@@ -50,6 +52,12 @@ def mock(server):
     yield httpmock
 
 
+@pytest.mark.flaky(
+    reruns=5,
+    rerun_delay=2,
+    condition="GITHUB_ACTIONS" in os.environ and platform.system() == "Darwin",
+    reason="Flaky; see iiasa/ixmp#489",
+)
 def test_check_single_model_access(mock, tmp_path, test_data_path):
     mock.when(
         "POST /access/list",

--- a/ixmp/tests/test_integration.py
+++ b/ixmp/tests/test_integration.py
@@ -1,4 +1,6 @@
 import logging
+import os
+import platform
 
 import numpy as np
 import pytest
@@ -93,6 +95,12 @@ def get_distance(scen):
     return scen.par("d").set_index(["i", "j"]).loc["san-diego", "topeka"]["value"]
 
 
+@pytest.mark.flaky(
+    reruns=5,
+    rerun_delay=2,
+    condition="GITHUB_ACTIONS" in os.environ and platform.system() == "Darwin",
+    reason="Flaky; see iiasa/ixmp#489",
+)
 def test_multi_db_run(tmpdir):
     # create a new instance of the transport problem and solve it
     mp1 = ixmp.Platform(backend="jdbc", driver="hsqldb", path=tmpdir / "mp1")

--- a/ixmp/tests/test_tutorials.py
+++ b/ixmp/tests/test_tutorials.py
@@ -1,4 +1,5 @@
 import os
+import platform
 import sys
 
 import numpy as np
@@ -6,7 +7,15 @@ import pytest
 
 from ixmp.testing import get_cell_output, run_notebook
 
+FLAKY = pytest.mark.flaky(
+    reruns=5,
+    rerun_delay=2,
+    condition="GITHUB_ACTIONS" in os.environ and platform.system() == "Darwin",
+    reason="Flaky; see iiasa/ixmp#489",
+)
 
+
+@FLAKY
 def test_py_transport(tutorial_path, tmp_path, tmp_env):
     fname = tutorial_path / "transport" / "py_transport.ipynb"
     nb, errors = run_notebook(fname, tmp_path, tmp_env)
@@ -16,6 +25,7 @@ def test_py_transport(tutorial_path, tmp_path, tmp_env):
     assert np.isclose(get_cell_output(nb, -5)["lvl"], 153.6750030517578)
 
 
+@FLAKY
 def test_py_transport_scenario(tutorial_path, tmp_path, tmp_env):
     fname = tutorial_path / "transport" / "py_transport_scenario.ipynb"
     nb, errors = run_notebook(fname, tmp_path, tmp_env)
@@ -25,6 +35,7 @@ def test_py_transport_scenario(tutorial_path, tmp_path, tmp_env):
     assert np.isclose(get_cell_output(nb, "scen-detroit-z")["lvl"], 161.324)
 
 
+@FLAKY
 @pytest.mark.rixmp
 # TODO investigate and resolve the cause of the time outs; remove this mark
 @pytest.mark.skipif(
@@ -36,6 +47,7 @@ def test_R_transport(tutorial_path, tmp_path, tmp_env):
     assert errors == []
 
 
+@FLAKY
 @pytest.mark.rixmp
 # TODO investigate and resolve the cause of the time outs; remove this mark
 @pytest.mark.skipif(

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -68,6 +68,7 @@ tests = [
   "pytest >= 5",
   "pytest-benchmark",
   "pytest-cov",
+  "pytest-rerunfailures",
 ]
 
 [project.scripts]

